xpdan/data_reduction.py

```diff
@@ -0,0 +1,382 @@
+
+from .glbl import an_glbl
+from .utils import _clean_info, _timestampstr
+
+from pyFAI.azimuthalIntegrator import AzimuthalIntegrator
+
+# top definition for minimal impacts on the code 
+if an_glbl._is_simulation:
+    db = MagicMock()
+    get_events = MagicMock()
+    get_images = MagicMock()
+else:
+    from databroker.databroker import get_table
+    from databroker.databroker import DataBroker as db
+    from databroker import get_images
+    from databroker import get_events
+
+w_dir = os.path.join(an_glbl.home, 'tiff_base')
+W_DIR = w_dir # in case of crashes in old codes
+
+
+
+class DataReduction:
+    """ class that handle operations on images from databroker header
+
+        Note: not a callback
+    """
+    def __init__(self, image_field=None):
+        # for file name 
+        self.fields = ['sample_name','sp_type', 'sp_requested_exposure']
+        self.labels = ['dark_frame']
+        self.data_fields = ['temperature']
+        self.root_dir_name = 'sample_name'
+        if image_field is None:
+            self.image_field = an_glbl.det_image_field
+
+    def _feature_gen(self, event):
+        ''' generate a human readable file name.
+
+        file name is generated by metadata information in event
+        run_start
+        '''
+        feature_list = []
+        run_start = event.descriptor['run_start']
+        uid = run_start['uid'][:6]
+        # get special label
+        for el in self.labels:
+                label = run_start.get(el, None)
+                if label is not None:
+                    feature_list.append(str(label))
+                else:
+                    pass
+        # get fields
+        for key in self.fields:
+            el = str(run_start.get(key, None))
+            if el is not None:
+                # truncate string length
+                if len(el) >12:
+                    value = el[:12]
+                # clear space
+                feature = _clean_info(el)
+                feature_list.append(feature)
+            else:
+                pass
+        # get data fields
+        for key in self.data_fields:
+            val = event['data'].get(key, None)
+            if el is not None:
+                feature = "{}={}".format(key, val)
+                feature_list.append(feature)
+            else:
+                pass
+        # get uid
+        feature_list.append(uid)
+        return "_".join(feature_list)
+
+    def pull_dark(self, header):
+        dark_uid = header.start.get(an_glbl.dark_field_key, None)
+        if dark_uid is None:
+            print("INFO: no dark frame is associated in this header, "
+                  "subrraction will not be processed")
+            return None
+        else:
+            dark_search = {'group': 'XPD', 'uid': dark_uid}
+            dark_header = db(**dark_search)
+            dark_img = np.asarray(get_images(dark_header,
+                                             self.image_field)).squeeze()
```

Comment:
due to plugins on area detector, we did have image with dimension equal to ``(1, 2048, 2048)``. I am not sure if that is the case now, but a squeeze doesn't seem to be harmful to me.